Refactor query command

bronzdoc opened the pull request:
# Description:
Attempt to make query command code clearer

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

deivid-rodriguez commented on lib/rubygems/commands/query_command.rb:
```diff
@@ -129,55 +137,55 @@ def display_header(type)
 
   #Guts of original execute
   def show_gems(name, prerelease)
```
You have extracted `prerelease` to a method, so this method argument is now unused.

deivid-rodriguez commented on lib/rubygems/commands/query_command.rb:
```diff
@@ -129,55 +137,55 @@ def display_header(type)
-    req = Gem::Requirement.default
-    # TODO: deprecate for real
-    dep = Gem::Deprecate.skip_during { Gem::Dependency.new name, req }
-    dep.prerelease = prerelease
```
I'm unsure about this deprecation but the current command is broken for local gems and the `--no-prerelease` option.

If we keep the above three lines and add `specs.select! { |spec| dep.match?(spec.name, spec.version) }` after the initial filtering of local specs, namely, where we do:

```ruby
specs = Gem::Specification.find_all do |s|
  s.name =~ name and req =~ s.version
end
```

Then I think that would fix that bug and we can remove the warning about it

```ruby
if prerelease? && !both?
  alert_warning("prereleases are always shown locally")
end
```

bronzdoc replied:
@deivid-rodriguez how can i reproduce the bug?

deivid-rodriguez replied:
```
$ ruby -Ilib bin/gem list bund

*** LOCAL GEMS ***

bundler (2.1.0.pre.1, 2.0.1, default: 1.17.3)

$ ruby -Ilib bin/gem query bund --prerelease
WARNING:  prereleases are always shown locally

*** LOCAL GEMS ***

bundler (2.1.0.pre.1, 2.0.1, default: 1.17.3)

$ ruby -Ilib bin/gem query bund --no-prerelease

*** LOCAL GEMS ***

bundler (2.1.0.pre.1, 2.0.1, default: 1.17.3)
```

If I pass the `--no-prerelease` parameter, prereleases should be excluded.

bronzdoc commented:
This is ready if you have any more feedback i appreciate it @deivid-rodriguez 👀  
I fixed the `--no-prerelease` bug and removed the warning, since is not needed anymore.

deivid-rodriguez commented on lib/rubygems/commands/query_command.rb:
```diff
@@ -123,6 +123,10 @@ def prerelease?
   end
 
+  def show_prereleases?
+    options[:prerelease].nil? || options[:prerelease]
```
You can reuse the `prerelease?` method here :)

bronzdoc replied:
how i missed that 😂 thanks!

bronzdoc commented:
> I would instead show them too by default for --remote too, but we can do that separately 
I'll do a followup PR after this is merged for that 👍

deivid-rodriguez commented on lib/rubygems/commands/query_command.rb:
```diff
@@ -78,46 +78,58 @@ def description # :nodoc:
-    names = Array(name)
-    names.each { |n| show_gems n, prerelease }
+  def show_prereleases?
+    options[:prerelease].nil? || prerelease?
```
You can use the `prerelease` method in the first part of the condition as well :)

bronzdoc replied:
isn't `prerelease?.nil?` to weird?

deivid-rodriguez replied:
Hahaha, you're right! Maybe we can call the method `prerelease` instead, since it's not strictly boolean anyways.

bronzdoc replied:
Fair 👍

bronzdoc commented:
@bundlerbot r+
